### PR TITLE
perms parameter cache, ratelimits and db destructure

### PR DESCRIPTION
**Pull request by fatchan**
1. cache parameters in permissions check to reduce created closures, since there are only 5 distinct permissions levels.
2. separate ratelimits from captchas, so I can make other actions contribute to ratelimits later.
3. make an index.js in db folder so I can require `.../db/` and destructure for easier requiring multiple db models.

### CODE REVIEW

File: helpers/checks/haspermsmiddleware.js [hunk 1]
```diff
@@ -1,8 +1,10 @@
 'use strict';
 
+const cache = {};
```
Comment (bilfr3d) [inline helpers/checks/haspermsmiddleware.js]: nice

File: db/ratelimits.js [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+module.exports = {
+
+	db,
+
+	resetQuota: (ip) => {
+		return db.deleteOne({ '_id': ip });
+	},
+
+	incrmentQuota: (ip, amount) => {
+		return db.findOneAndUpdate(
```
Comment (fatchan) [inline db/ratelimits.js]: Indent this properly 3Head

File: models/pages/captcha.js [hunk 3]
```diff
@@ -1,15 +1,15 @@
...
 	, generateCaptcha = require(__dirname+'/../../helpers/captcha/captchagenerate.js');
 
 module.exports = async (req, res, next) => {
 
 	let captchaId;
 	try {
-		const ratelimit = await Captchas.incrmentQuota(res.locals.ip);
-		if (ratelimit > 12) { // 12 per minute = 1 per 5 seconds within a minute (with burst)
-			return res.status(429).redirect(`/img/ratelimit.png`);
+		const ratelimit = await Ratelimits.incrmentQuota(res.locals.ip, 10);
```
Comment (fatchan) [inline models/pages/captcha.js]: Was making the ratelimit for captchas effectively 10/min instead of 12/min intentional?

Comment (bilfr3d) [reply models/pages/captcha.js]: u r big dum dum

**Comment by bilfr3d** [comment]
Great work

**Pull request merged**